node/actors/network/src/gossip/mod.rs

```diff
@@ -148,37 +161,91 @@ impl Network {
         .await;
     }
 
-    /// Task that keeps hearing about new votes and looks for an L1 batch qc.
-    /// It will propagate the QC if there's enough votes.
+    /// Task that reacts to new votes being added and looks for an L1 batch QC.
+    /// It persists the certificate once the quorum threshold is passed.
     pub(crate) async fn run_batch_qc_finder(&self, ctx: &ctx::Ctx) -> ctx::Result<()> {
         let Some(attesters) = self.genesis().attesters.as_ref() else {
+            tracing::info!("no attesters in genesis, not looking for batch QCs");
             return Ok(());
         };
         let genesis = self.genesis().hash();
-        let mut sub = self.batch_votes.subscribe();
+
+        let mut recv_votes = self.batch_votes.subscribe();
+        let mut recv_status = self.attestation_status.subscribe();
+
+        let mut prev_batch_number = None;
+
+        'status: loop {
+            // Wait until the status indicates that we're ready to sign the next batch.
+            // This is not strictly necessary but avoids repeatedly finding the same quorum, or having to skip it until it changes.
+            let next_batch_number =
+                sync::wait_for_some(ctx, &mut recv_status, |s| match s.next_batch_to_attest {
+                    next if next == prev_batch_number => None,
+                    next => next,
+                })
+                .await?;
+
+            // Next time we'll look for something new.
+            prev_batch_number = Some(next_batch_number);
+
+            // Get rid of all previous votes. We don't expect this to go backwards without regenesis, which will involve a restart.
+            self.batch_votes
+                .set_min_batch_number(next_batch_number)
+                .await;
+
+            // Now wait until we find the next quorum, whatever it is:
+            // * on the main node, if attesters are honest, they will vote on the next batch number and the main node will not see gaps
+            // * on external nodes the votes might be affected by changes in the value returned by the API, and there might be gaps
+            // What is important, though, is that the batch number does not move backwards while we look for a quorum, because attesters
+            // (re)casting earlier votes will go ignored by those fixed on a higher min_batch_number, and gossip will only be attempted once.
+            // The possibility of this will be fixed by deterministally picking a start batch number based on fork indicated by genesis.
+            loop {
+                let quorum_opt = {
+                    let votes = sync::changed(ctx, &mut recv_votes).await?;
```

Comment:
Are you sure you don't need "recv_votes.mark_changed()" before this await? In particular `subscribe()` returns a receiver which is *seen*.

Reply:
Hm, I wouldn't want to call it here because then it would become a tight loop. I didn't know `.subscribe()` returns it seen, that's good to keep in mind, however in this case it doesn't matter because of the call to `set_min_batch_number` which happens before this loop, which will cause the whole `BatchVotes` to change, so this call will get something immediately, and then every time a vote is added.